### PR TITLE
validation, refactor: remove CChainState::GetMempool()

### PR DESCRIPTION
**Pull request by jonatack**
Simplify CChainState validation code by removing an unnecessary GetMempool() getter.

Reason: A trivial getter or setter adds no semantic value; unless I am missing something, the data item (CChainState::m_mempool) could just as well be public.

Relevant C++ Core Guidelines (among others)
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c131-avoid-trivial-getters-and-setters
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-public
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected

This change allows simplifying the tx_pool fuzzer chainstate code in the second commit.

### CODE REVIEW

File: src/test/fuzz/tx_pool.cpp [hunk 1]
```diff
@@ -143,7 +143,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
     CTxMemPool tx_pool_{/*estimator=*/nullptr, /*check_ratio=*/1};
     MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(&tx_pool_);
 
-    chainstate.SetMempool(&tx_pool);
+    chainstate.SetMempool(tx_pool);
```
Comment (lsilva01) [inline src/test/fuzz/tx_pool.cpp:146]: `m_mempool` can be accessed directly.

```suggestion
    chainstate.m_mempool = &tx_pool;
```

**Review by lsilva01 — APPROVED** [review]
tACK c1038a6 with some nits

File: src/test/fuzz/tx_pool.cpp [hunk 2]
```diff
@@ -32,9 +32,9 @@ struct MockedTxPool : public CTxMemPool {
 class DummyChainState final : public CChainState
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/test/fuzz/tx_pool.cpp]: nice! done

File: src/test/fuzz/tx_pool.cpp (hunk 1 above)
Comment (jonatack) [reply src/test/fuzz/tx_pool.cpp]: thanks, done

**Comment by jonatack** [comment]
Thank you @lsilva01, I've added a commit authored by you for the fuzz code simplifications.

**Review by lsilva01 — APPROVED** [review]
reACK b86b91a

**Review by shaavan — APPROVED** [review]
crACK b86b91a0d30f8bb487366b2d2f8aba9b5e8741f1

**Comment by jnewbery** [comment]
I'm ~0 on this. I think a crucial difference between what is being proposed here and the advice in https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c131-avoid-trivial-getters-and-setters is that here there was only a getter and no setter. Prior to this PR, only `CChainState` or a derived class could update `m_mempool`. Now, any client code can modify it. That's a change in the class's contract with its clients.

**Comment by jonatack** [comment]
Hm. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-public might be relevant.

**Comment by MarcoFalke** [comment]
> Hm. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-public might be relevant. If we want it to be encapsulated, then it should be private. Public and protected are the same category.

A reference is a non-nullable "const" pointer (For pointers there are two dimensions of `const`), so I don't think comparing m_mempool with m_blockman (or other references) is fair.

**Comment by jonatack** [comment]
My thought isn't about comparison with the other data members, but rather that m_mempool would preferably be either public (with less and simpler code) as done here, or private with a setter if it is intended to be encapsulated.

**Comment by MarcoFalke** [comment]
My reply to that would be: https://github.com/bitcoin/bitcoin/pull/23465#discussion_r760832977

**Comment by jonatack** [comment]
After reading all the early review in #23465 I see that all of these options were previously proposed by @lsilva01; sorry about that.

If `m_mempool` should be encapsulated, I agree with @jnewbery and with @promag in  https://github.com/bitcoin/bitcoin/pull/23465#discussion_r756697995.

Agree with avoiding test-only helpers. In this case there may be good counter-arguments:

- making it `protected` instead of `private` appears to be only for the fuzz test as well (and isn't explicit as to why)
- it is less safe and more complex to maintain than `private`
- it seems simpler and less error-prone to grep the codebase for a setter than to hunt for workarounds (and `m_mempool` is used frequently, and there is also a `BlockAssembler::m_mempool` private data member) or than by manually making it `private` and then compiling to see what fails

However, since I now see that @lsilva01 did propose for it to be private in https://github.com/bitcoin/bitcoin/commit/358f94a28176f766634d9701f1688abfcf761c28, I'm not going to update this pull to do it, but may support such a change in future review where makes sense.

**Pull request closed without merging**